### PR TITLE
Fix single value columns for ECM classifier

### LINKED ISSUE #75: ECM Classification Binary Label Issue; Deadline Sensitive
Hello there, I'm running into a bit of an issue trying to use an ECMClassifier object's fit function with a vector pandas.DataFrame object. I've tried looking around but I've had no success as all the solutions provided do not relate to this recordlinkage library.

To help with the understanding of the problem, I'll describe what the dataframes I am working with are like (I can't show actual pictures of them due to work privacy).
Dataframe1 contains seven columns in the following order: Block_Key, cust_name, physical_padress1, physical_address2, city, state, zip.
Dataframe2 has eight columns: Block_Key, business_name, tradestyle, sec_trdstl, physical_street, physical_city, physical_state, physical_zip.

They are indexed with blocking using the Block_Key to return a list of MultiIndex. 

    def block_key_pairing(self, block_keys):
        df1Data = pandas.read_excel(self.Dataframe1)
        df2Data = pandas.read_excel(self.Dataframe2)
        candidate_df = pandas.DataFrame()
        pair_list = []
        for block_key in block_keys:
            bkIndexer = BlockKeyIndexer(block_key)
            pair = bkIndexer.index(df1Data, df2Data)
            for index in pair:
                candidate_df.loc[index] = block_key
            pair_list.append(pair)
        candidate_df.to_excel('C:\\Users\\Documents\\Pairs & Block Keys.xlsx')
        return pair_list

BlockKeyIndexer is a class object for custom indexing due to the regular Indexer returning pairs with different block keys because they are different by one or two characters.
The function that the candidate pairs are used is shown below, and it also where I fiddled with classifications over the course of a couple days.

    def comparator(self, candidate_links):
        comp = recordlinkage.Compare()
        vector_df = pandas.DataFrame()

        comp.string("cust_name", "business_name", method='levenshtein',label='Business Name')
        comp.string("cust_name", "tradestyle", method='levenshtein',label='Trade Style')
        comp.string("cust_name", "sec_trdstl", method='levenshtein',label='Secondary TS')
        comp.string("physical_address_1", "physical_street", method='levenshtein',label='Primary Address')
        comp.string("physical_address_2", "physical_street", method='levenshtein',label='Secondary Address')
        comp.string("city", "physical_city", method='levenshtein',label='City')
        comp.string("state", "physical_state", method='levenshtein',label='State')
        comp.numeric("zip", "physical_zip",label='Zip')

        df1Data = pandas.read_excel(self.Dataframe1)
        df2Data = pandas.read_excel(self.Dataframe2)

        for candidate_pair in candidate_links:
            vector_df = vector_df.append(comp.compute(candidate_pair, cdmData, dnbData))
           
        vector_df=vector_df.astype(np.int_)
        ecm=recordlinkage.ECMClassifier(init='jaro',binarize=0.8)
        result_ecm=ecm.fit(vector_df)
        print(len(result_ecm))

        return vector_df

So a vector dataframe is created called vector_df and when trying to use the fit function yields this error

    ValueError: Only binary labels are allowed for 'jaro'method. Column 2 has 1 different labels.

I've become very perplexed by this because it seems to imply that Columns 0 & 1 are acceptable.
So I figured the problem was because the values inside the vector_df weren't binary, so I tried modifying the comparator function like so.

    def comparator(self, candidate_links):
        comp = recordlinkage.Compare()

        vector_df = pandas.DataFrame()

        comp.string("cust_name", "business_name", method='levenshtein',label='Business Name')
        comp.string("cust_name", "tradestyle", method='levenshtein',label='Trade Style')
        #comp.string("cust_name", "sec_trdstl", method='levenshtein',label='Secondary TS')
        comp.string("physical_address_1", "physical_street", method='levenshtein',label='Primary Address')
        comp.string("physical_address_2", "physical_street", method='levenshtein',label='Secondary Address')
        comp.string("city", "physical_city", method='levenshtein',label='City')
        comp.string("state", "physical_state", method='levenshtein',label='State')
        comp.numeric("zip", "physical_zip",label='Zip')

        df1Data = pandas.read_excel(self.Dataframe1)
        df2Data = pandas.read_excel(self.Dataframe2)

        for candidate_pair in candidate_links:
            vector_df = vector_df.append(comp.compute(candidate_pair,df1Data,df2Data))
           
        vector_df=vector_df.astype(np.int_)
        ecm=recordlinkage.ECMClassifier(init='jaro',binarize=0.8)
        result_ecm=ecm.fit(vector_df)
        print(len(result_ecm))

        return vector_df

However this yields a new error when running.

    ValueError: could not broadcast input array from shape (12) into shape (13)

I've tried debugging and I looked inside the library to see what the process is but I still can't figure out what I need to do to the vector_df to have my code run correctly.

I am new to the recordlinkage library, having only worked with it for three weeks now.

### PR DESCRIPTION
**Pull request by J535D165**
This Pull Request solves some (not all) issues with the ECM classifier. Single value columns result in errors at the moment (see #75). 

This PR relates to #100 and #101 as well. I'm not sure how yet. I'm investigating this. 

Closes #75.

### CODE REVIEW

File: recordlinkage/algorithms/nb_sklearn.py [hunk 1]
```diff
@@ -211,8 +211,12 @@ def _fit_data(self, X):
             # initialise binarizer and save
             binarizer = LabelBinarizer()
 
+            if self.binarize:
+                binarizer.classes_ = numpy.array([0, 1])
```
Comment (jpweytjens) [inline recordlinkage/algorithms/nb_sklearn.py]: Numpy is imported as np.

Comment (J535D165) [reply recordlinkage/algorithms/nb_sklearn.py]: This is a very good catch. I fixed it and added a test.

**Comment by J535D165** [comment]
I'm a bit struggling with the convergence issues you pointed out. Your save log is nice but results sometimes in probabilities larger than 1. And sometimes I'm facing a RuntimeWarning: `divide by zero encountered in log`.  But I think this is the moment to stop iterating a throw a warning that explains that there is a convergence issue. Any remarks? Btw, your PR is very useful in debugging this. Many thanks for this proposed solution.

**Comment by jpweytjens** [comment]
I have found a solution for the convergence issue that doesn't require the `safe_log` function. 

Convergence requires that `class_log_prior_close` and `feature_log_prob_close` are both `True`. As noted in #100, the joint log likelihood can have elements that decrease with every iteration. This in turn causes the logarithm to return infinity when these elements finally become 0. 

For values close to zero, the derivative of the logarithm is very steep such that `feature_log_prob_close` can never be `True`. Small changes in `feature_prob_` are greatly amplified in `feature_log_prob_`, despite `feature_prob_` not changing more than the given `atol`.

A solution consists of changing

```
feature_log_prob_close = np.allclose(
    feature_log_prob_,
    self.feature_log_prob_,
    atol=self.atol
)
```
to
```
feature_log_prob_close = np.allclose(
    feature_prob_,
    np.exp(self.feature_log_prob_),
    atol=self.atol
)
```

I tested this with the `binary_vectors` function and had good estimates for multiple datasets that didn't converge with the previous convergence criterion. The number of estimated matches are always close to the used `match_percentage`.

While this seems to work, I don't understand why the joint log likelihood can decrease with every iteration for some datasets. This shouldn't be the case I think.

**Comment by J535D165** [comment]
Thanks Johannes, good find. I think it was a bad idea to use the log probs for the stopping condition. I made a commit that implements your solution. It seems to work well. I will discuss the decrease in the joint log-likelihood with a statistician. Hopefully, that will make things clear.

**Pull request merged**